diff --git a/library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsChunkSource.java b/library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsChunkSource.java
index 74362abdb0..7f488a6769 100644
--- a/library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsChunkSource.java
+++ b/library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsChunkSource.java
@@ -470,15 +470,19 @@ import org.checkerframework.checker.nullness.qual.MonotonicNonNull;
     @Nullable
     Uri initSegmentKeyUri =
         getFullEncryptionKeyUri(playlist, segmentBaseHolder.segmentBase.initializationSegment);
-    out.chunk = maybeCreateEncryptionChunkFor(initSegmentKeyUri, selectedTrackIndex);
-    if (out.chunk != null) {
-      return;
+    if (! (mediaDataSource instanceof HlsDecryptingDataSource)) {
+      out.chunk = maybeCreateEncryptionChunkFor(initSegmentKeyUri, selectedTrackIndex);
+      if (out.chunk != null) {
+        return;
+      }
     }
     @Nullable
     Uri mediaSegmentKeyUri = getFullEncryptionKeyUri(playlist, segmentBaseHolder.segmentBase);
-    out.chunk = maybeCreateEncryptionChunkFor(mediaSegmentKeyUri, selectedTrackIndex);
-    if (out.chunk != null) {
-      return;
+    if (! (mediaDataSource instanceof HlsDecryptingDataSource)) {
+      out.chunk = maybeCreateEncryptionChunkFor(mediaSegmentKeyUri, selectedTrackIndex);
+      if (out.chunk != null) {
+        return;
+      }
     }
 
     boolean shouldSpliceIn =
@@ -507,6 +511,7 @@ import org.checkerframework.checker.nullness.qual.MonotonicNonNull;
             isTimestampMaster,
             timestampAdjusterProvider,
             previous,
+            mediaSegmentKeyUri,
             /* mediaSegmentKey= */ keyCache.get(mediaSegmentKeyUri),
             /* initSegmentKey= */ keyCache.get(initSegmentKeyUri),
             shouldSpliceIn);
